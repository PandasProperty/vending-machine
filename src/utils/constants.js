export const VENDING_MACHINE_CONFIG = {
    ROWS: 7,
    COLUMNS: 5,
    TOTAL_ITEMS: 15
};

export const VENDING_MACHINE_STATES = {
    IDLE: 'IDLE',
    READING: 'READING',
    PAYMENT: 'PAYMENT',
    DELIVER: 'DELIVER'
};

export const VENDING_MACHINE_CURRENCY = 'RON';

export const VENDING_MACHINE_MONEY = {
    0: {
        value: 0.5,
        label: '50 bani'
    },
    1: {
        value: 1,
        label: '1 leu'
    },
    2: {
        value: 5,
        label: '5 lei'
    },
    3: {
        value: 10,
        label: '10 lei'
    }
};