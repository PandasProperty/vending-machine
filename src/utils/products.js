
import { random, reduce, map, find } from 'lodash';
import {
    faHamburger,
    faPizzaSlice,
    faEgg,
    faCheese,
    faBreadSlice,
    faBacon,
    faHotdog,
    faIceCream,
    faFish,
    faCookie,
    faCandyCane,
    faCarrot,
    faLemon,
    faAppleAlt,
    faGlassWhiskey,
    faGlassMartini,
    faWineBottle,
    faBeer,
    faCoffee,
    faMugHot
} from '@fortawesome/free-solid-svg-icons';
import { VENDING_MACHINE_CONFIG } from './constants';

const colors = ['green', 'yellow', 'red', 'blue', 'purple', 'orange'];

const availableProducts = [{
    icon: faHamburger,
    name: 'Hamburger',
    price: 15
}, {
    icon: faPizzaSlice,
    name: 'Pizza',
    price: 15
}, {
    icon: faEgg,
    name: 'Kinder Egg',
    price: 5
}, {
    icon: faCheese,
    name: 'Cheese',
    price: 15
}, {
    icon: faBreadSlice,
    name: 'Toast',
    price: 2
}, {
    icon: faBacon,
    name: 'Bacon',
    price: 5
}, {
    icon: faHotdog,
    name: 'Hotdog',
    price: 7
}, {
    icon: faIceCream,
    name: 'Ice Cream',
    price: 5
}, {
    icon: faFish,
    name: 'Fish Chips',
    price: 15
}, {
    icon: faCookie,
    name: 'Cookie',
    price: 5
}, {
    icon: faCandyCane,
    name: 'Candy',
    price: 3
}, {
    icon: faCarrot,
    name: 'Carrot',
    price: 2
}, {
    icon: faLemon,
    name: 'Lemon',
    price: 3
}, {
    icon: faAppleAlt,
    name: 'Apple',
    price: 3
}, {
    icon: faGlassWhiskey,
    name: 'Whiskey',
    price: 20
}, {
    icon: faGlassMartini,
    name: 'Martini',
    price: 20
}, {
    icon: faWineBottle,
    name: 'Wine Bottle',
    price: 35
}, {
    icon: faBeer,
    name: 'Beer',
    price: 10
}, {
    icon: faCoffee,
    name: 'Coffee',
    price: 5
}, {
    icon: faMugHot,
    name: 'Hot Chocolate',
    price: 5
}];

export const generateVendingMachineContent = () => {
    const vendingMachineProducts = {};
    for (let rowIndex = 0; rowIndex < VENDING_MACHINE_CONFIG.ROWS; rowIndex++) {
        for (let columnIndex = 0; columnIndex < VENDING_MACHINE_CONFIG.COLUMNS; columnIndex++) {
            const productIndex = random(0, availableProducts.length - 1);
            const productId = rowIndex * VENDING_MACHINE_CONFIG.COLUMNS + columnIndex;
            vendingMachineProducts[productId] = {
                id: productId,
                ...availableProducts[productIndex],
                color: colors[random(0, colors.length - 1)],
                quantity: random(1, VENDING_MACHINE_CONFIG.TOTAL_ITEMS)
            };
        };
    };
    return vendingMachineProducts;
};

const computeCoins = (coins, length, ammount, max) => {
    if (!ammount) {
        return {
            coins,
            nrOfCoins: 0
        };
    }
    let rest = {
        coins,
        nrOfCoins: max
    };
    for (let i=0; i<length; i++) {
        if (coins[i].value <= ammount && coins[i].quantity > 0) {
            
            const newCoins = map(coins, coin => ({
                ...coin,
                quantity: coin.id === coins[i].id ? coin.quantity - 1 : coin.quantity
            }));

            let subRest = computeCoins(newCoins, length, ammount - coins[i].value, max);

            if (subRest.nrOfCoins !== max && subRest.nrOfCoins + 1 < rest.nrOfCoins) {
                rest = {
                    ...subRest,
                    nrOfCoins: subRest.nrOfCoins + 1
                };
                break;
            }
        }
    }
    return rest;
};

export const getMaximumChange = (coins, ammount) => {
    let max = 0;
    map(coins, coin => max += coin.quantity);
    return computeCoins(coins, coins.length, ammount, max);
};

export const getMachineCapital = (coins) => (
    reduce(coins, (sum, coin) => sum + coin.quantity * coin.value, 0)
);

export const getCoinsDescription = (oldCoins, newCoins) => {
    const description = [];
    map(oldCoins, oldCoin => {
        const newCoin = find(newCoins, newCoin => oldCoin.id === newCoin.id);
        description.push({
            ...newCoin,
            quantity: oldCoin.quantity - newCoin.quantity
        })
    });
    return description;
};
