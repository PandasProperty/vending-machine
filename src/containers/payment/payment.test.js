import React from 'react';
import { Provider } from 'react-redux';
import { map } from 'lodash';
import { faMugHot } from '@fortawesome/free-solid-svg-icons';
import Payment from '.';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { VENDING_MACHINE_STATES, VENDING_MACHINE_CONFIG } from '../../utils/constants';

const mockStore = configureStore([]);

describe('Vending Machine', () => {
    const store = mockStore({
        vendingMachine: {
            products: map(
                Array(VENDING_MACHINE_CONFIG.ROWS * VENDING_MACHINE_CONFIG.COLUMNS), (item, index) => ({
                    color: 'orange',
                    icon: faMugHot,
                    id: index,
                    name: 'Cheese',
                    price: 15,
                    quantity: 7,
                })
            ),
            inputProduct: 1,
            mode: VENDING_MACHINE_STATES.DELIVER
        },
        errorMessage: 'Error message',
        userBalance: 40,
        userChange: null
    });
    
    it('Payment componet', () => {
        const component = renderer.create(
            <Provider store={store}>
                <Payment />
            </Provider>
        );
        
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});
