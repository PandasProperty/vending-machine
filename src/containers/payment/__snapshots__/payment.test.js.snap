// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Vending Machine Payment componet 1`] = `
<div
  className="payment"
>
  <label
    className="payment__info"
  >
    The vending maching support only 
    50 bani, 1 leu, 5 lei, 10 lei
    .
  </label>
  <div
    className="display"
  >
    <div
      className="display__label"
    >
      Balance
    </div>
    <div
      className="display__value"
    >
      undefined RON
    </div>
  </div>
  <div
    className="payment__payment-input"
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-registered fa-w-16 fa-2x payment__icon"
      data-icon="registered"
      data-prefix="fas"
      focusable="false"
      onClick={[Function]}
      role="img"
      style={Object {}}
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M285.363 207.475c0 18.6-9.831 28.431-28.431 28.431h-29.876v-56.14h23.378c28.668 0 34.929 8.773 34.929 27.709zM504 256c0 136.967-111.033 248-248 248S8 392.967 8 256 119.033 8 256 8s248 111.033 248 248zM363.411 360.414c-46.729-84.825-43.299-78.636-44.702-80.98 23.432-15.172 37.945-42.979 37.945-74.486 0-54.244-31.5-89.252-105.498-89.252h-70.667c-13.255 0-24 10.745-24 24V372c0 13.255 10.745 24 24 24h22.567c13.255 0 24-10.745 24-24v-71.663h25.556l44.129 82.937a24.001 24.001 0 0 0 21.188 12.727h24.464c18.261-.001 29.829-19.591 21.018-35.587z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
    <div
      className="payment__info"
    >
      Press the 'R' button for rest.
    </div>
  </div>
  <div
    className="error__container"
  >
    
  </div>
  <div
    className="payment__wallet"
  >
    <div
      className="payment__info"
    >
      Your Wallet
    </div>
    <div
      className="payment__coins"
    >
      <div
        className="payment__coin"
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-coins fa-w-16 fa-2x payment__icon"
          data-icon="coins"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 405.3V448c0 35.3 86 64 192 64s192-28.7 192-64v-42.7C342.7 434.4 267.2 448 192 448S41.3 434.4 0 405.3zM320 128c106 0 192-28.7 192-64S426 0 320 0 128 28.7 128 64s86 64 192 64zM0 300.4V352c0 35.3 86 64 192 64s192-28.7 192-64v-51.6c-41.3 34-116.9 51.6-192 51.6S41.3 334.4 0 300.4zm416 11c57.3-11.1 96-31.7 96-55.4v-42.7c-23.2 16.4-57.3 27.6-96 34.5v63.6zM192 160C86 160 0 195.8 0 240s86 80 192 80 192-35.8 192-80-86-80-192-80zm219.3 56.3c60-10.8 100.7-32 100.7-56.3v-42.7c-35.5 25.1-96.5 38.6-160.7 41.8 29.5 14.3 51.2 33.5 60 57.2z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
        <label>
          50 bani
        </label>
      </div>
      <div
        className="payment__coin"
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-coins fa-w-16 fa-2x payment__icon"
          data-icon="coins"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 405.3V448c0 35.3 86 64 192 64s192-28.7 192-64v-42.7C342.7 434.4 267.2 448 192 448S41.3 434.4 0 405.3zM320 128c106 0 192-28.7 192-64S426 0 320 0 128 28.7 128 64s86 64 192 64zM0 300.4V352c0 35.3 86 64 192 64s192-28.7 192-64v-51.6c-41.3 34-116.9 51.6-192 51.6S41.3 334.4 0 300.4zm416 11c57.3-11.1 96-31.7 96-55.4v-42.7c-23.2 16.4-57.3 27.6-96 34.5v63.6zM192 160C86 160 0 195.8 0 240s86 80 192 80 192-35.8 192-80-86-80-192-80zm219.3 56.3c60-10.8 100.7-32 100.7-56.3v-42.7c-35.5 25.1-96.5 38.6-160.7 41.8 29.5 14.3 51.2 33.5 60 57.2z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
        <label>
          1 leu
        </label>
      </div>
      <div
        className="payment__coin"
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-coins fa-w-16 fa-2x payment__icon"
          data-icon="coins"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 405.3V448c0 35.3 86 64 192 64s192-28.7 192-64v-42.7C342.7 434.4 267.2 448 192 448S41.3 434.4 0 405.3zM320 128c106 0 192-28.7 192-64S426 0 320 0 128 28.7 128 64s86 64 192 64zM0 300.4V352c0 35.3 86 64 192 64s192-28.7 192-64v-51.6c-41.3 34-116.9 51.6-192 51.6S41.3 334.4 0 300.4zm416 11c57.3-11.1 96-31.7 96-55.4v-42.7c-23.2 16.4-57.3 27.6-96 34.5v63.6zM192 160C86 160 0 195.8 0 240s86 80 192 80 192-35.8 192-80-86-80-192-80zm219.3 56.3c60-10.8 100.7-32 100.7-56.3v-42.7c-35.5 25.1-96.5 38.6-160.7 41.8 29.5 14.3 51.2 33.5 60 57.2z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
        <label>
          5 lei
        </label>
      </div>
      <div
        className="payment__coin"
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-coins fa-w-16 fa-2x payment__icon"
          data-icon="coins"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 405.3V448c0 35.3 86 64 192 64s192-28.7 192-64v-42.7C342.7 434.4 267.2 448 192 448S41.3 434.4 0 405.3zM320 128c106 0 192-28.7 192-64S426 0 320 0 128 28.7 128 64s86 64 192 64zM0 300.4V352c0 35.3 86 64 192 64s192-28.7 192-64v-51.6c-41.3 34-116.9 51.6-192 51.6S41.3 334.4 0 300.4zm416 11c57.3-11.1 96-31.7 96-55.4v-42.7c-23.2 16.4-57.3 27.6-96 34.5v63.6zM192 160C86 160 0 195.8 0 240s86 80 192 80 192-35.8 192-80-86-80-192-80zm219.3 56.3c60-10.8 100.7-32 100.7-56.3v-42.7c-35.5 25.1-96.5 38.6-160.7 41.8 29.5 14.3 51.2 33.5 60 57.2z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
        <label>
          10 lei
        </label>
      </div>
    </div>
  </div>
</div>
`;
